docs/design/core/metrics/prototype/MetricPublisher.java

```diff
@@ -64,5 +54,5 @@
      * Implementations can also call publish method for every reported metric. But this can be expensive and
      * is not recommended.
      */
-    CompletableFuture<Void> publish();
+    CompletableFuture<Void> publish(MetricEvents metricEvents);
```

Comment:
Although the events themselves are immutable, the MetricEvents registry is not. This brings up questions in my mind like what happens if a metric is added during the publish cycle, is that undefined behavior? What are they meant to do with the MetricEvents object once it's been published, discard it? I think we need either better documentation explaining all that or a small rethink around this aspect of it.

Reply:
Yeah the mutability is an issue. I think that can be solved by passing around a `MetricEvents.Builder` instead, and `MetricEvents` becomes immutable and publish can take consumer builder. WDYT?

Reply:
Well two different approaches come to my mind:
1) Make the MetricEvents a closeable object, and once it is closed no new metrics can be added and only closed MetricEvents can be published. This requires orchestration upstream though to manage these multiple MetricEvents objects.
2) Make the whole thing a stream that has publishers attached to it with markers so they know the last event successfully published.

Your idea with passing around the builder really conforms more to my first approach.

Reply:
I think the builder and suggestion 1 approach are the cleanest IMO. requiring a `MetricEvents` rather than a `MetricEvents.Builder` makes the "closing" explicit so I'm sort of leaning toward this approach

Reply:
Is it not the job of the aggregators to aggregate and bundle metrics to then be passed as a batch to publisher? Maybe we're missing a whole layer here : Metrics flow as a stream to aggregators, aggregators create metric bundles and send those to publishers, publishers publish the bundle. It's then up to the aggregator to handle the stream to batch orchestration.

Reply:
Discussed this offline. The interface doesn't require implementations of `MetricPublisher` to immediately publish the metrics given in the `publish` call and allows for batching/aggregation on the back end; the aggregator is collapsed into the publisher. This falls in line with the concept of streams (in this case a streams of `MetricEvents`) mentioned above.